lib/smart_proxy_host_reports/ansible_processor.rb

```diff
@@ -82,46 +74,56 @@ def spool_report
 
   private
 
-  def log_level(log)
-    if log["failed"]
-      "err"
-    elsif log["changed"]
-      "notice"
-    else
-      "info"
+  def process_keywords(result)
+    if result["failed"]
+      add_keywords("HasFailure", "AnsibleTaskFailed:#{result["task"]["action"]}")
+    elsif result["changed"]
+      add_keywords("HasChange")
     end
   end
 
-  def log_source(log)
-    # Temporary until  is resolved
-    log["module"]
+  def process_level(result)
+    if result["failed"]
+      result["level"] = "err"
+    elsif result["changed"]
+      result["level"] = "notice"
+    else
+      result["level"] = "info"
+    end
   end
 
-  def log_message(log)
-    return log["msg"] if log["failed"]
+  def process_message(result)
+    msg = "N/A"
+    return msg if result["task"].nil? || result["task"]["action"].nil?
+    return result["result"]["msg"] if result["failed"]
+    result_tree = result["result"]
+    task_tree = result["task"]
+    raise("Report do not contain required 'results/result' element") unless result_tree
+    raise("Report do not contain required 'results/task' element") unless task_tree
+    module_args_tree = result_tree.dig("invocation", "module_args")
 
-    case log["module"]
-    when "package"
-      log["results"].empty? ? log["msg"] : log["results"]
-    when "template"
-      module_args = log["invocation"]["module_args"]
-      "Rendered template #{module_args["_original_basename"]} to #{log["dest"]}"
-    when "service"
-      "Service #{log["name"]} #{log["state"]} (enabled: #{log["enabled"]})"
-    when "group"
-      "User group #{log["name"]} #{log["state"]}, gid: #{log["gid"]}"
-    when "user"
-      "User #{log["name"]} #{log["state"]}, uid: #{log["uid"]}"
-    when "cron"
-      module_args = log["invocation"]["module_args"]
-      "Cron job: #{module_args["minute"]} #{module_args["hour"]} #{module_args["day"]} #{module_args["month"]} #{module_args["weekday"]} #{module_args["job"]} (disabled: #{module_args["disabled"]})"
-    when "copy"
-      module_args = log["invocation"]["module_args"]
-      "Copy #{module_args["_original_basename"]} to #{log["dest"]}"
+    case task_tree["action"]
+    when "ansible.builtin.package"
```

Comment:
```suggestion
    when "package"
```

Let's stick with `foreman_ansible` approach. `ansible.builtin.module` appears if use ansible via command line and use built-in modules. For custom roles/tasks it's missing.

Reply:
I am not sure what you mean, I just tested this with:

```
ANSIBLE_LOAD_CALLBACK_PLUGINS=1 sudo ansible localhost -vvv -m ansible.builtin.package -a "name=vim state=present"

{
  "format": "ansible",
  "id": "73985b36-7a1e-4175-8acc-4b3630113c7a",
  "host": "report.example.com",
  "proxy": "localhost",
  "reported_at": "2021-10-08T07:15:49.341191",
  "results": [
    {
      "result": {
        "msg": "Nothing to do",
        "changed": false,
        "rc": 0,
        "invocation": {
          "module_args": {
            "name": [
              "vim"
            ],
            "state": "present",
            "allow_downgrade": false,
            "autoremove": false,
            "bugfix": false,
            "disable_gpg_check": false,
            "download_only": false,
            "installroot": "/",
            "install_repoquery": true,
            "install_weak_deps": true,
            "security": false,
            "skip_broken": false,
            "update_cache": false,
            "update_only": false,
            "validate_certs": true,
            "lock_timeout": 30
          }
        },
        "_ansible_no_log": false
      },
      "task": {
        "action": "ansible.builtin.package",
        "async_val": 0,
        "async": 0,
        "delay": 5,
        "poll": 15,
        "retries": 3,
        "name": "",
        "connection": "smart",
        "check_mode": false,
        "diff": false,
        "any_errors_fatal": false,
        "throttle": 0,
        "become": false,
        "become_method": "sudo",
        "uuid": "1c697a0b-ae03-a393-94fb-000000000009",
        "finalized": true,
        "squashed": true
      },
      "failed": false,
      "level": "info",
      "friendly_message": "Package(s) vim are present: Nothing to do"
    }
  ],
  "keywords": [

  ],
  "telemetry": {
    "parse": 0.03382004797458649,
    "process": 0.015477067790925503
  },
  "status": {
    "applied": 0,
    "failed": 0,
    "skipped": 0
  },
  "check_mode": false
}
```

Reply:
Yeah, exactly: you have here `["task"]["action"] == "ansible.builtin.package"` the action is the same you've used to run ansible `-m ansible.builtin.package`. But when you run a playbook you will have `["task"]["action"] == "package"`. See last line:
![ScreenShot-1633693643852](https://user-images.githubusercontent.com/32508194/136551633-58a3b9d4-08e1-4382-9516-fbbf9b3de838.png)

Reply:
This and all below is still not resolved...

Reply:
No, I do not have `"action": "package"`, see here and in other example dumps:

https://github.com/theforeman/smart_proxy_host_reports/pull/5/files#diff-d22ea879447ac1084da738dca64815331e0ae8b1d72e4ef8151144b59f4a7c8bR51

Maybe you had an old version of the callback?

I am not sure about splitting it by dot, what is added benefit? Looks like Ansible uses some kind of dot-syntax for module names, so let's just use it.